platform/x11/cog-platform-x11.c

```diff
@@ -0,0 +1,896 @@
+        { -1,  1 }, { 1,  1 },
+        { -1, -1 }, { 1, -1 },
+    };
+    static const float texture_coords[4][2] = {
+        { 0, 0 }, { 1, 0 },
+        { 0, 1 }, { 1, 1 },
+    };
+
+    eglMakeCurrent (egl_data.display, egl_data.surface, egl_data.surface, egl_data.context);
+
+    glViewport (0, 0, xcb_data.width, xcb_data.height);
+    glClearColor (1, 1, 1, 1);
+    glClear (GL_COLOR_BUFFER_BIT);
+
+    glUseProgram (gl_data.program);
+
+    glActiveTexture (GL_TEXTURE0);
+    glBindTexture (GL_TEXTURE_2D, gl_data.texture);
+    egl_data.image_target_texture (GL_TEXTURE_2D,
+                                   wpe_fdo_egl_exported_image_get_egl_image (image));
+    glUniform1i (gl_data.uniform_texture, 0);
+
+    glVertexAttribPointer (gl_data.attr_pos, 2, GL_FLOAT, GL_FALSE, 0, position_coords);
+    glVertexAttribPointer (gl_data.attr_texture, 2, GL_FLOAT, GL_FALSE, 0, texture_coords);
+
+    glEnableVertexAttribArray (gl_data.attr_pos);
+    glEnableVertexAttribArray (gl_data.attr_texture);
+
+    glDrawArrays (GL_TRIANGLE_STRIP, 0, 4);
+
+    glDisableVertexAttribArray (gl_data.attr_pos);
+    glDisableVertexAttribArray (gl_data.attr_texture);
+
+    eglSwapBuffers (egl_data.display, egl_data.surface);
+
+    wpe_view_data.image = image;
+    xcb_data.needs_initial_paint = false;
+    xcb_data.needs_frame_completion = true;
+    xcb_schedule_repaint ();
+}
+
+static void
+xcb_frame_completion (void)
+{
+    if (wpe_view_data.image) {
+        wpe_view_backend_exportable_fdo_egl_dispatch_release_exported_image (wpe_host_data.exportable, wpe_view_data.image);
+        wpe_view_data.image = NULL;
+    }
+
+    wpe_view_backend_exportable_fdo_dispatch_frame_complete (wpe_host_data.exportable);
+}
+
+static void
+xcb_handle_key_press (xcb_key_press_event_t *event)
+{
+    uint32_t keysym = xkb_state_key_get_one_sym (xkb_data.state, event->detail);
+    uint32_t unicode = xkb_state_key_get_utf32 (xkb_data.state, event->detail);
+
+    struct wpe_input_keyboard_event input_event = {
+        event->time,
+        keysym,
+        unicode,
+        true,
+        xkb_data.modifiers
+    };
+    wpe_view_backend_dispatch_keyboard_event (wpe_view_data.backend, &input_event);
+}
+
+static void
+xcb_handle_key_release (xcb_key_press_event_t *event)
+{
+    uint32_t keysym = xkb_state_key_get_one_sym (xkb_data.state, event->detail);
+    uint32_t unicode = xkb_state_key_get_utf32 (xkb_data.state, event->detail);
+
+    struct wpe_input_keyboard_event input_event = {
+        event->time,
+        keysym,
+        unicode,
+        false,
+        xkb_data.modifiers
+    };
+    wpe_view_backend_dispatch_keyboard_event (wpe_view_data.backend, &input_event);
+}
+
+static void
+xcb_handle_axis (xcb_button_press_event_t *event, int16_t axis_delta[2])
+{
+#if HAVE_2D_AXIS_EVENT
+    struct wpe_input_axis_2d_event input_event = { 0, };
+    input_event.base.type = wpe_input_axis_event_type_mask_2d | wpe_input_axis_event_type_motion_smooth;
```

Comment:
This could use a designated initializer, which would guarantee that members not specified will be initialized with zeroes, too.